File: c++/hash_table/q249.cpp
class Solution {
public:
    vector<vector<string>> groupStrings(vector<string>& strings) {
        unordered_map<string,vector<string>> um;
        vector<vector<string>> res;
        for(int i = 0; i < strings.size(); i++){
            um[shift(strings[i])].push_back(strings[i]);
        }
        for(unordered_map<string,vector<string>>::iterator it = um.begin(); it != um.end(); it++){
            vector<string> unit = it->second;
            sort(unit.begin(),unit.end());
            res.push_back(unit);
        }

        return res;
    }
private:
    string shift(string& s){
        string t = "";
        for(int i = 1; i < s.size(); i++){
            int diff = s[i] - s[i-1];
            if(diff < 0)
                diff += 26;
            t += ('a' + diff + ',');
        }
        return t;
    }
};

File: python/hash_table/q249.py
"""
Given a string, we can "shift" each of its letter to its successive letter, for example: "abc" -> "bcd". We can keep "shifting" which forms the sequence:

"abc" -> "bcd" -> ... -> "xyz"
Given a list of strings which contains only lowercase alphabets, group all strings that belong to the same shifting sequence.

For example, given: ["abc", "bcd", "acef", "xyz", "az", "ba", "a", "z"],
A solution is:

[
  ["abc","bcd","xyz"],
  ["az","ba"],
  ["acef"],
  ["a","z"]
]
"""
class Solution(object):
    def groupStrings(self, strings):
        """
        :type strings: List[str]
        :rtype: List[List[str]]
        """
        ret = []
        dic = {}
        for val in strings:
            dic[self.shift(val)] = dic.get(self.shift(val),[])
            dic[self.shift(val)].append(val)

        for li in dic.values():
            li.sort()
            ret.append(li)

        return ret

    def shift(self,s):
        temp = ""
        for i,l in enumerate(s[1:]):
            diff = ord(s[i]) - ord(s[i-1])
            if diff < 0:
                diff += 26
            temp += (chr(ord('a')+diff)+",")
        return temp
